src/SBBox.cpp

```diff
@@ -283,9 +327,238 @@ namespace galsim {
         dbg<<"Box shoot: N = "<<N<<std::endl;
         dbg<<"Target flux = "<<getFlux()<<std::endl;
         boost::shared_ptr<PhotonArray> result(new PhotonArray(N));
-        for (int i=0; i<result->size(); i++)
-            result->setPhoton(i, _width*(u()-0.5), _height*(u()-0.5), _flux/N);
+        double fluxPerPhoton = _flux/N;
+        for (int i=0; i<N; i++)
+            result->setPhoton(i, _width*(u()-0.5), _height*(u()-0.5), fluxPerPhoton);
         dbg<<"Box Realized flux = "<<result->getTotalFlux()<<std::endl;
         return result;
     }
+
+
+
+    SBTopHat::SBTopHat(double radius, double flux, const GSParamsPtr& gsparams) :
+        SBProfile(new SBTopHatImpl(radius,flux,gsparams)) {}
+
+    SBTopHat::SBTopHat(const SBTopHat& rhs) : SBProfile(rhs) {}
+
+    SBTopHat::~SBTopHat() {}
+
+    double SBTopHat::getRadius() const 
+    {
+        assert(dynamic_cast<const SBTopHatImpl*>(_pimpl.get()));
+        return static_cast<const SBTopHatImpl&>(*_pimpl).getRadius(); 
+    }
+
+    SBTopHat::SBTopHatImpl::SBTopHatImpl(double radius, double flux,
+                                         const GSParamsPtr& gsparams) :
+        SBProfileImpl(gsparams),
+        _r0(radius), _r0sq(_r0*_r0), _flux(flux),
+        _norm(_flux / (M_PI * _r0sq))
+    {
+    }
+
+
+    double SBTopHat::SBTopHatImpl::xValue(const Position<double>& p) const 
+    {
+        double rsq = p.x*p.x + p.y*p.y;
+        if (rsq < _r0sq) return _norm;
+        else return 0.;
+    }
+
+    std::complex<double> SBTopHat::SBTopHatImpl::kValue(const Position<double>& k) const
+    {
+        double kr0sq = (k.x*k.x + k.y*k.y) * _r0sq;
+        return kValue2(kr0sq);
+    }
+
+    std::complex<double> SBTopHat::SBTopHatImpl::kValue2(double kr0sq) const
+    {
+        if (kr0sq < 1.e-4) {
+            // Error from omitting next term is about 1.e-16 for kr0sq = 1.e-4
+            return _flux * (1. - kr0sq * ( (1./8.) + (1./192.) * kr0sq ));
```

Comment:
Is this formula supposed to be so obvious (and the size of next term so obvious) that it doesn't require an explanation?  It didn't seem so obvious to me, so I would appreciate a comment describing it if you don't mind.